README
Packages required: 
 * camlp4
 * camlp4-extra
 * ocaml-findlib

To make: 
$ make all

You will be asked for the root password, as installation is part of
the make process.  This because some of the packages need, in order to
be built, for previous packages to be already installed. 

You can find in the various subdirectories Makefiles that are
perfectly functional, should you wish to build the packages and
install them one by one in the traditional way.



sexplib/CHANGES
2008-04-24:  Made Sexp-interface manifest.

2008-03-20:  Fixed META-file (missing num dependency).

2008-03-17:  Improved META-file.

2008-03-13:  Fully allow function types in converters.  Raise runtime
             exceptions on converting from S-expressions instead when
             function type encountered.

2008-02-11:  Fixed code generation problems with variance annotations
             in signatures, and empty types.

2007-12-17:  Added support for generating signature entries for
             S-expression converters.  Thanks to Till Varoquaux
             <till.varoquaux@gmail.com> for the patch!

2007-11-29:  Added support for converting big_int, nat, num, and ratio.

2007-11-26:  Added support for parsing from bigstrings (char bigarrays).

2007-11-02:  Added syntax support for option types to use the ordinary
             sum type syntax.  This should improve readability.
             The old syntax will be accepted, too, if
             Conv.read_old_option_format is set to true (this is
             currently the default).  The old format will be used for
             writing if Conv.write_old_option_format is true (currently
             the default).  The old syntax is deprecated and will
             probably not be supported by default in the near future.
             Reading new-style option values will always succeed.

2007-09-14:  Fixed bug in S-expression preprocessor concerning record
             field names.

2007-08-06:  Added support for converting functions to S-expressions.

2007-07-20:  Fixed position information and improved speed of S-expression
             parser.  Fixed S-expression macro bug concerning contained
             polymorphic variants.

2007-06-28:  Improved Sexplib code generation.

2007-06-22:  Fixed escaping bug in S-expression parser.

2007-06-01:  Added correct handling of recursive types + test case.

2007-04-18:  Added missing conversion functions from S-expressions to
             pairs and triples.

2007-03-21:  Updated OCamlMakefile.

2007-03-02:  Improved error messages when parsing illegal type definitions.

2007-01-30:  Added triple conversions.

2006-11-22:  Updated OCamlMakefile.

2006-10-13:  Improved checking of records for extra or duplicate fields.

2006-09-06:  Added support for polymorphic record fields.

2006-09-05:  Added support for manifest types.

2006-08-16:  Improved error messages.

2006-07-28:  Added a new, hand-written S-expression parser that supports
             partial parsing and should be approx. 10x faster than the
             previous one.

2006-06-20:  Fixed a code generation problem leading to compilation
             errors concerning the use of type aliases within polymorphic
             variant type definitions.

             This fix also solves potential erroneous appearances of
             backtracking exceptions in user code.

2006-03-21:  Added -for-pack option to Makefile and cleaned up
             distribution for a new public release.

2006-03-13:  Sexplib now accepts capitalized booleans.

2006-03-03:  Added customizable indentation levels.
             Improved documentation.
             Fixed API-problem concerning backward compatibility.

2006-03-03:  Added customizable indentation levels.
             Improved documentation.

2006-03-01:  Added a missing flush for string conversions with a buffer.

2006-02-08:  Eliminated unused variable warnings in Sexplib-generated code.

2006-01-11:  Added functions for pretty-printing to buffers.  Improved
             performance of outputting S-expressions to channels.

2006-01-09:  Added functions load_sexp and load_sexps.

2006-01-04:  Changed float conversion from %E to %G (more readable).

2005-12-28:  Made machine representation for S-expressions more compact

2005-12-15:  Fixed a problem appearing with OCaml-release 3.08.4:  CamlP4
             obviously performs more strict checking on some constructs
             now and crashed with an exception when generating
             S-expression code for records containing only one field
             ("singleton tuple problem").  This problem is fixed now.

2005-11-25:  Fixed problem with type variables that could not be generalized.

2005-11-23:  Added a missing case in type definitions (path alias)

2005-11-17:  Major release: 2.0

             Fixed a major design problem.  The user now has to pass lex
             buffers instead of channels to input-functions.  Reason:
             trailing characters in channels were lost due to ocamllex
             buffering them in the non-exposed lex buffer.  This lex
             buffer is now exposed.  The functions have been renamed
             ("input_X" -> "scan_X") to reflect this change.

2005-11-16:  Added label to conversion function "input_cnv_sexps".

2005-11-11:  Fixed a bug in the pretty-printer: strings in atoms were
             not escaped in the function "to_string_mach" (and therefore
             also "to_string").

2005-11-07:  Initial release.


vec.ml
(* Vec: implementation of extensible arrays. 

   Copyright Luca de Alfaro <lda@dealfaro.org>, 2007.
   All rights reserved.  

   Version 1.3

   Based on Xavier Leroy's code for Set and Map.

   This file is distributed under the terms of the GNU Library General
   Public License, with the special exception on linking described in
   file LICENSE.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version,
   with the following special exception:

   You may link, statically or dynamically, a "work that uses the
   Library" with a publicly distributed version of the Library to
   produce an executable file containing portions of the Library, and
   distribute that executable file under terms of your choice, without
   any of the additional requirements listed in clause 6 of the GNU
   Library General Public License.  By "a publicly distributed version
   of the Library", we mean either the unmodified Library as
   distributed by INRIA, or a modified version of the Library that is
   distributed under the conditions defined in clause 2 of the GNU
   Library General Public License.  This exception does not however
   invalidate any other reasons why the executable file might be
   covered by the GNU Library General Public License.

   This library is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   The GNU Library General Public License is available at
   http://www.gnu.org/copyleft/lgpl.html; to obtain it, you can also
   write to the Free Software Foundation, Inc., 59 Temple Place -
   Suite 330, Boston, MA 02111-1307, USA.
 *)

type 'a t =
    Empty
      (* meaning: node of left el, left n.els, element, right el, right n. els, heigth *)
  | Node of 'a t * int * 'a * 'a t * int * int 

exception Vec_index_out_of_bounds

let height = function
    Empty -> 0
  | Node(_,_,_,_,_,h) -> h

let length = function 
    Empty -> 0 
  | Node (_, cl, _, _, cr, _) -> 1 + cl + cr 

let makenode l d r =
  let (hl, cl) = match l with Empty -> (0,0) | Node(_,lcl,_,_,lcr,h) -> (h, lcl + lcr + 1) in 
  let (hr, cr) = match r with Empty -> (0,0) | Node(_,rcl,_,_,rcr,h) -> (h, rcl + rcr + 1) in 
  Node(l, cl, d, r, cr, (if hl >= hr then hl + 1 else hr + 1))

let rec create d = function 
    0 -> Empty 
  | n ->
      let ml = n / 2 in 
      let mr = n - ml - 1 in 
      makenode (create d ml) d (create d mr)

(* bal assumes that l and r are of similar height *)
let bal l d r =
  let hl = match l with Empty -> 0 | Node(_,_,_,_,_,h) -> h in
  let hr = match r with Empty -> 0 | Node(_,_,_,_,_,h) -> h in
  if hl > hr + 2 then begin
    match l with
      Empty -> invalid_arg "Vec.bal"
    | Node(ll, _, ld, lr, _, _) ->
        if height ll >= height lr then
          makenode ll ld (makenode lr d r)
        else begin
          match lr with
            Empty -> invalid_arg "Vec.bal"
          | Node(lrl, _, lrd, lrr, _, _) ->
              makenode (makenode ll ld lrl) lrd (makenode lrr d r)
        end
  end else if hr > hl + 2 then begin
    match r with
      Empty -> invalid_arg "Vec.bal"
    | Node(rl, _, rd, rr, _, _) ->
        if height rr >= height rl then
          makenode (makenode l d rl) rd rr
        else begin
          match rl with
            Empty -> invalid_arg "Vec.bal"
          | Node(rll, _, rld, rlr, _, _) ->
              makenode (makenode l d rll) rld (makenode rlr rd rr)
        end
  end 
  else makenode l d r 
      
(* This is a recursive version of balance, which balances a tree all the way down. 
   The trees l and r can be of any height, but they need to be internally balanced.  
   Useful to implement concat. *)
let rec recbal l d r =
  let hl = match l with Empty -> 0 | Node(_,_,_,_,_,h) -> h in
  let hr = match r with Empty -> 0 | Node(_,_,_,_,_,h) -> h in
  if hl > hr + 2 then begin
    match l with
      Empty -> invalid_arg "Vec.bal"
    | Node(ll, _, ld, lr, _, _) ->
        if height ll >= height lr then
          makenode ll ld (recbal lr d r)
        else begin
          match lr with
            Empty -> invalid_arg "Vec.bal"
          | Node(lrl, _, lrd, lrr, _, _) ->
              makenode (makenode ll ld lrl) lrd (recbal lrr d r)
        end
  end else if hr > hl + 2 then begin
    match r with
      Empty -> invalid_arg "Vec.bal"
    | Node(rl, _, rd, rr, _, _) ->
        if height rr >= height rl then
          makenode (recbal l d rl) rd rr
        else begin
          match rl with
            Empty -> invalid_arg "Vec.bal"
          | Node(rll, _, rld, rlr, _, _) ->
              makenode (recbal l d rll) rld (makenode rlr rd rr)
        end
  end 
  else makenode l d r 
      
let empty = Empty
  
let is_empty = function Empty -> true | _ -> false

let singleton d = Node (Empty, 0, d, Empty, 0, 1)

let rec get i = function 
    Empty -> raise Vec_index_out_of_bounds
  | Node (l, cl, d, r, cr, _) -> 
      if i < cl then get i l 
      else if i > cl then get (i - cl - 1) r 
      else d 

let rec set i d = function 
    Empty -> raise Vec_index_out_of_bounds
  | Node (l, cl, dd, r, cr, _) -> 
      if i < cl then makenode (set i d l) dd r  
      else if i > cl then makenode l dd (set (i - cl - 1) d r)
      else makenode l d r 

let rec append d = function 
    Empty -> Node (Empty, 0, d, Empty, 0, 1)
  | Node (l, _, dd, r, _, _) -> 
      bal l dd (append d r)
  
let setappend d0 d i v =
  let l = length v in 
  if l > i then set i d v 
  else begin
    let vr = ref v in 
    for j = l to i - 1 do 
      vr := append d0 !vr 
    done; 
    append d !vr
  end 

let rec leftmost = function
    Empty -> raise Vec_index_out_of_bounds
  | Node(Empty, _, d, r, _, _) -> d
  | Node(l, _, d, r, _, _) -> leftmost l
      
let rec remove_leftmost = function
    Empty -> invalid_arg "Vec.remove_min_elt"
  | Node(Empty, _, d, r, _, _) -> r
  | Node(l, _, d, r, _, _) -> bal (remove_leftmost l) d r
      
(* merge uses bal, not recbal, so it only works for trees of similar height *)
let merge t1 t2 =
  match (t1, t2) with
    (Empty, t) -> t
  | (t, Empty) -> t
  | (_, _) ->
      let d = leftmost t2 in
      bal t1 d (remove_leftmost t2)
	
(* Concat works also for t1 and t2 of very different heights *)
let concat t1 t2 =
  match (t1, t2) with
    (Empty, t) -> t
  | (t, Empty) -> t
  | (_, _) ->
      let d = leftmost t2 in
      recbal t1 d (remove_leftmost t2)
	
let rec pop i = function
    Empty -> raise Vec_index_out_of_bounds
  | Node(l, cl, d, r, cr, h) ->
      if i < cl then 
	let (e, v) = pop i l in 
	(e, bal v d r)
      else if i > cl then 
	let (e, v) = pop (i - cl - 1) r in 
	(e, bal l d v)
      else (d, merge l r)

let rec remove i = function 
    Empty -> raise Vec_index_out_of_bounds
  | Node(l, cl, d, r, cr, h) ->
      if i < cl then 
	bal (remove i l) d r 
      else if i > cl then 
	bal l d (remove (i - cl - 1) r)
      else merge l r 

	  
let rec insert i d = function 
    Empty -> begin
      if i = 0 
      then Node (Empty, 0, d, Empty, 0, 1)
      else raise Vec_index_out_of_bounds
    end
  | Node(l, cl, dd, r, cr, h) -> 
      if i < cl then bal (insert i d l) dd r 
      else if i > cl then bal l dd (insert (i - cl - 1) d r)
      else bal l d (insert 0 dd r)

let rec sub i j v = match v with 
    Empty -> Empty
  | Node (l, cl, dd, r, cr, _) -> 
      if i >= j then Empty
	(* Important for sharing *)
      else if i <= 0 && j >= cl + cr + 1 then v 
      else begin 
	if j <= cl then sub i j l 
	else if j = cl + 1 then append dd (sub i cl l)
	else if i = cl then insert 0 dd (sub 0 (j - cl - 1) r)
	else if i > cl then sub (i - cl - 1) (j - cl - 1) r
	else begin
	  (* dd straddles the interval *)
	  let ll = sub i cl l in 
	  let rr = sub 0 (j - cl - 1) r in 
	  recbal ll dd rr 
	end
      end

let rec iter f = function
    Empty -> ()
  | Node(l, _, d, r, _, _) ->
      iter f l; f d; iter f r

let rec iteri f v = 
  let rec offsetiteri k = function
    Empty -> ()
  | Node(l, cl, d, r, _, _) ->
      offsetiteri k l; f (k + cl) d; offsetiteri (k + cl + 1) r
  in offsetiteri 0 v

let rec reviter f = function
    Empty -> ()
  | Node(l, _, d, r, _, _) ->
      iter f r; f d; iter f l

let rec rangeiter f i j = function
    Empty -> ()
  | Node(l, cl, d, r, cr, _) ->
      if i < j then begin 
	if i < cl && j > 0 then rangeiter f i j l; 
	if i <= cl && j > cl then f d;
	if j > cl + 1 && i <= cl + cr + 1 then rangeiter f (i - cl - 1) (j - cl - 1) r
      end

let rec revrangeiter f i j = function
    Empty -> ()
  | Node(l, cl, d, r, cr, _) ->
      if i < j then begin 
	if j > cl + 1 && i <= cl + cr + 1 then rangeiter f (i - cl - 1) (j - cl - 1) r;
	if i <= cl && j > cl then f d;
	if i < cl && j > 0 then rangeiter f i j l
      end

let rangeiteri f i j v = 
  let rec offsetrangeiteri k i j = function
      Empty -> ()
    | Node(l, cl, d, r, cr, _) ->
      if i < j then begin 
	if i < cl && j > 0 then offsetrangeiteri k i j l; 
	if i <= cl && j > cl then f (k + cl) d;
	if j > cl + 1 && i <= cl + cr + 1 
	then offsetrangeiteri (k + cl + 1) (i - cl - 1) (j - cl - 1) r
      end
  in offsetrangeiteri 0 i j v 

let revrangeiteri f i j v = 
  let rec offsetrevrangeiteri k i j = function
      Empty -> ()
    | Node(l, cl, d, r, cr, _) ->
      if i < j then begin 
	if j > cl + 1 && i <= cl + cr + 1 
	then offsetrevrangeiteri (k + cl + 1) (i - cl - 1) (j - cl - 1) r;
	if i <= cl && j > cl then f (k + cl) d;
	if i < cl && j > 0 then offsetrevrangeiteri k i j l
      end
  in offsetrevrangeiteri 0 i j v 

let rec map f = function
    Empty -> Empty
  | Node(l, cl, d, r, cr, h) -> Node(map f l, cl, f d, map f r, cr, h)
      
let mapi f v = 
  let rec offsetmapi k = function 
      Empty -> Empty
    | Node(l, cl, d, r, cr, h) -> 
	Node(offsetmapi k l, cl, f (k + cl) d, offsetmapi (k + cl + 1) r, cr, h)
  in offsetmapi 0 v 

let rec fold f v accu =
  match v with
    Empty -> accu
  | Node(l, _, d, r, _, _) ->
      fold f r (f d (fold f l accu))
	
let foldi f v accu =
  let rec offsetfoldi k v accu = 
    match v with
      Empty -> accu
    | Node(l, cl, d, r, _, _) ->
	offsetfoldi (k + cl + 1) r (f (k + cl) d (offsetfoldi k l accu))
  in offsetfoldi 0 v accu
	
let rangefoldi f i j v accu = 
  let rec offsetrangefoldi k i j v accu = 
    match v with 
      Empty -> accu
    | Node (l, cl, d, r, cr, _) -> 
	if i >= j then accu
	else begin 
	  let al = if i < cl && j > 0 then offsetrangefoldi k i j l accu else accu in 
	  let ad = if i <= cl && j > cl then f (cl + k) d al else al in 
	  if j > cl + 1 && i <= cl + cr + 1
	  then offsetrangefoldi (k + cl + 1) (i - cl - 1) (j - cl - 1) r ad
	  else ad
	end
  in offsetrangefoldi 0 i j v accu 

let revfoldi f v accu =
  let rec offsetrevfoldi k v accu = 
    match v with
      Empty -> accu
    | Node(l, cl, d, r, _, _) ->
	offsetrevfoldi k l (f (k + cl) d (offsetrevfoldi (k + cl + 1) r accu))
  in offsetrevfoldi 0 v accu

let revrangefoldi f i j v accu = 
  let rec offsetrevrangefoldi k i j v accu = 
    match v with 
      Empty -> accu
    | Node (l, cl, d, r, cr, _) -> 
	if i >= j then accu
	else begin 
	  let ar = if j > cl + 1 && i <= cl + cr + 1
	  then offsetrevrangefoldi (k + cl + 1) (i - cl - 1) (j - cl - 1) r accu
	  else accu
	  in 
	  let ad = if i <= cl && j > cl then f (cl + k) d ar else ar in 
	  if i < cl && j > 0 then offsetrevrangefoldi k i j l ad else ad
	end
  in offsetrevrangefoldi 0 i j v accu;; 

let rec exists p = function 
    Empty -> false
  | Node (l, _, d, r, _, _) -> 
      exists p l || p d || exists p r;;

let find p n v = 
  (* k is how many elements there are to the left; 
     i is where to start the search *)
  let rec offset_find k i = function 
    Empty -> None 
  | Node (l, cl, d, r, _, _) -> begin 
      match offset_find k i l with 
	Some x -> Some x 
      | None -> begin 
	  if k <= i && p d 
	  then Some (cl + k, d) 
	  else offset_find (cl + k + 1) (i - cl - 1) r
	end
    end
  in offset_find 0 n v 

let rec filter p = function 
    Empty -> Empty 
  | Node (l, _, d, r, _, _) -> 
      let lf = filter p l in 
      let lr = filter p r in 
      if p d then recbal lf d lr else concat lf lr;;

let rec of_list = function 
    [] -> Empty 
  | d :: l -> insert 0 d (of_list l)

let to_list v = 
  let rec auxtolist accu = function 
      Empty -> accu 
    | Node (l, _, d, r, _, _) -> auxtolist (d :: auxtolist accu r) l 
  in auxtolist [] v;;

let rec to_array w = 
  match w with 
    Empty -> [||]
  | Node (l, cl, d, r, cr, _) -> 
      begin 
	(* Creates the array *)
	let n = cl + cr + 1 in 
	let a = Array.make n d in 
	(* and fills it *)
	let rec fill a k = function 
	    Empty -> a
	  | Node (l, cl, d, r, _, _) -> begin
	      ignore (fill a k l); 
	      Array.set a (k + cl) d; 
	      fill a (k + cl + 1) r
	    end
	in fill a 0 w 
      end
	      
let of_array a =
  let f accu el = append el accu in 
  Array.fold_left f Empty a;; 

(* Visitor paradigm *)

(* Post-order visitor *)
let visit_post ve vn a = 
  let rec f = function
      Empty -> ve
    | Node (l, _, d, r, _, _) -> 
	let rl = f l in 
	let rr = f r in 
	vn rl d rr
  in f a;;

(* In-order visitor *)
let visit_in ve vl vr a = 
  let rec f = function 
      Empty -> ve
    | Node (l, _, d, r, _, _) ->
	let rl = vl (f l) d in 
	vr rl (f r)
  in f a;; 

(* Unit testing *)

if false then begin 
  let print_vec v = 
    let p i = Printf.printf " %d" i in 
    Printf.printf "["; iter p v; Printf.printf " ]\n"
  in
  let v = of_list [0; 1; 2; 3; 4; 5; 6; 7; 8; 9; 10; 11; 12; 13; 14; 15; 16; 17; 18; 19] in 
  print_vec v; 

  let print_l (b: unit) (d: int) : unit = 
    begin print_int d; print_string " " end in 
  let print_r (c: unit) (b: unit) : unit = () in 
  visit_in () print_l print_r v;
  print_string "\n"; 

  print_vec (of_array (to_array v));
  Printf.printf "%d\n" (get 5 v); 
  let u = concat v v in
  Printf.printf "%d\n" (length u); 
  Printf.printf "\n"; 
  print_vec (sub 4 9 (set 5 100 v)); 
  print_vec (sub (-10) 43 v); 
  print_vec (sub 30 30 v); 
  print_vec (sub 1 11 v); 
  let mult2 x = 2 * x in 
  print_vec (map mult2 (sub 3 8 v));
  let ff i d acc = (string_of_int i) ^ ":" ^ (string_of_int d) ^ ", " ^ acc in 
  Printf.printf "%s\n" (revrangefoldi ff 5 13 v "");
  print_vec (setappend (-2) (-1) 16 (sub 4 12 v))

end;;
